### ISSUE TITLE
copy.copy(sentinel.foo) is not sentinel.foo

### ISSUE THREAD

**Issue by GoogleCodeExporter** [testing-cabal/mock#244]
```
What steps will reproduce the problem?
1. import mock, copy
2. copy.copy(sentinel.foo) is sentinel.foo -> False

What is the expected output? What do you see instead?
I would expect sentinels to not be copied; they are meant to be singletons.

What version of the product are you using? On what operating system?
Ubuntu 14.10, python-mock 1.0.1-3
I also tested it on trunk.

Please provide any additional information below.
In MAAS we have the following snippet in an early import to solve the problem:

  # Make sure that sentinel objects are not copied.
  sentinel_type = type(mock.sentinel.foo)
  copy._copy_dispatch[sentinel_type] = copy._copy_immutable
  copy._deepcopy_dispatch[sentinel_type] = copy._copy_immutable

```

Original issue reported on code.google.com by `gavinpan...@gmail.com` on 22 Jan 2015 at 11:33

**Comment by rbtcollins** [comment]
Huh? This neds some more background I think. Could be related to the possible fix for #245

**Comment by allenap** [comment]
I noticed this behaviour when creating a couple of test helpers that we use with `Mock.side_effect`:

```python
def always_succeed_with(result):
    """Return a callable that always returns a successful Deferred.

    The callable allows (and ignores) all arguments, and returns a shallow
    `copy` of `result`.
    """
    def always_succeed(*args, **kwargs):
        return defer.succeed(copy(result))
    return always_succeed

def always_fail_with(result):
    """Return a callable that always returns a failed Deferred.

    The callable allows (and ignores) all arguments, and returns a shallow
    `copy` of `result`.
    """
    def always_fail(*args, **kwargs):
        return defer.fail(copy(result))
    return always_fail
```

To be honest we'd be fine without the copy behaviour. Still, `sentinel.foo is copy(sentinel.foo)` and `sentinel.bar is deepcopy(sentinel.bar)` should hold for pragmatic reasons: it's fairly confusing to be told by a test that `sentinel.baz is not sentinel.baz`. There might also be notions of correctness — singletons should be singletons, say — but I'm not so bothered by those :)

**Comment by rbtcollins** [comment]
Thanks @allenap - so I think this should be fixed, and its an upstream bug, so if you could file it in bugs.python.org that would be great. Thanks!

**Comment by katzenbaer** [comment]
@rbtcollins Can you explain why this is an upstream bug?

**Comment by allenap** [comment]
I think this bug can be fixed with some `copyreg` fiddling which could/should be done within `mock` itself:

```python
import copy
import copyreg
from unittest.mock import sentinel

_sentinel_type = type(sentinel.foo)

def _sentinel_construct(name):
    return getattr(sentinel, name)

def _sentinel_pickle(s):
    return _sentinel_construct, (s.name, )

copyreg.pickle(_sentinel_type, _sentinel_pickle)

assert copy.copy(sentinel.foo) is sentinel.foo
assert copy.deepcopy(sentinel.foo) is sentinel.foo
```

This would be preferable to the meddling with `copy._copy_dispatch` and `copy._deepcopy_dispatch` from my original bug report.

**Comment by katzenbaer** [comment]
@rbtcollins Can we do this? I verified that @allenap's solution also works on 2.7.10 by importing `copy_reg` instead of `copyreg`.

**Comment by rbtcollins** [comment]
Its an upstream bug because it also has to be done in the python3 unittest/mock.py codebase

**Comment by allenap** [comment]
@rbtcollins Sorry Rob, I didn't act on your earlier suggestion to report this upstream. I don't know if I missed it, or I procrastinated and forgot, or something else, but... there has been [discussion of this upstream](https://bugs.python.org/issue20804) and [a fix](http://hg.python.org/lookup/98f061402fcf) was landed a few weeks ago. It looks like this will only appear in 3.7 though; a separate [documentation fix](http://hg.python.org/lookup/398b73dbb1a0) was landed to 3.5 that states:

> The ``sentinel`` attributes don't preserve their identity when they are copied or pickled.

(I assume this will make its way into 3.6 too?)

Unfortunately, this means that doing a ``copyreg`` / ``copy_reg`` fix would contradict the documentation in Python 3.5. We could encapsulate it in a ``make_sentinel_copyable`` function for convenience, but perhaps at this point it's not worth pursuing further, and instead we should trust that people who are tripped up by this behaviour will find their way to this issue or upstream's.

**Comment by cjw296** [comment]
Hmm, okay, it's unclear if these guys need backporting:
http://hg.python.org/lookup/98f061402fcf
http://hg.python.org/lookup/398b73dbb1a0
...so closing this now.

If there is stuff that needs backporting, a link to git revs in https://github.com/python/cpython that need to come across would be good!
